src/OrchardCore.Modules/OrchardCore.Autoroute/Services/AutorouteEntries.cs

```diff
@@ -1,92 +1,148 @@
+                {
+                    document.Paths[entry.ContainedContentItemId] = entry;
+                }
+                else
+                {
+                    document.Paths[entry.ContentItemId] = entry;
                 }
             }
         }
 
-        public void RemoveEntries(IEnumerable<AutorouteEntry> entries)
+        public void RemoveEntries(AutorouteDocument document, IEnumerable<AutorouteEntry> entries)
         {
-            lock (this)
+            foreach (var entry in entries)
             {
-                foreach (var entry in entries)
+                // Evict all entries related to a container item from autoroute entries.
+                var entriesToRemove = document.Paths.Values.Where(x => x.ContentItemId == entry.ContentItemId && !String.IsNullOrEmpty(x.ContainedContentItemId));
+                foreach (var entryToRemove in entriesToRemove)
                 {
-                    // Evict all entries related to a container item from autoroute entries.
-                    var entriesToRemove = _paths.Values.Where(x => x.ContentItemId == entry.ContentItemId && !String.IsNullOrEmpty(x.ContainedContentItemId));
-                    foreach (var entryToRemove in entriesToRemove)
-                    {
-                        _paths.Remove(entryToRemove.ContainedContentItemId);
-                        _contentItemIds.Remove(entryToRemove.Path);
-                    }
-
-                    _paths.Remove(entry.ContentItemId);
-                    _contentItemIds.Remove(entry.Path);
+                    document.Paths.Remove(entryToRemove.ContainedContentItemId);
+                    document.ContentItemIds.Remove(entryToRemove.Path);
                 }
+
+                document.Paths.Remove(entry.ContentItemId);
+                document.ContentItemIds.Remove(entry.Path);
             }
         }
+
+        /// <summary>
+        /// Loads the autoroute document for updating and that should not be cached.
+        /// </summary>
+        private Task<AutorouteDocument> LoadDocumentAsync() => DocumentManager.GetMutableAsync(CreateDocumentAsync);
+
+        /// <summary>
+        /// Gets the autoroute document for sharing and that should not be updated.
+        /// </summary>
+        private Task<AutorouteDocument> GetDocumentAsync() => DocumentManager.GetImmutableAsync(CreateDocumentAsync);
+
+        private async Task<AutorouteDocument> CreateDocumentAsync()
+        {
+            var indexes = await Session.QueryIndex<AutoroutePartIndex>(i => i.Published).ListAsync();
+            var entries = indexes.Select(i => new AutorouteEntry(i.ContentItemId, i.Path, i.ContainedContentItemId, i.JsonPath));
+
+            var document = new AutorouteDocument();
+            AddEntries(document, entries);
+
+            return document;
+        }
+
+        private static ISession Session => ShellScope.Services.GetRequiredService<ISession>();
```

Comment:
Is there a way we can avoid having to make these static? It is, I imagine, the reason you are having to have special methods, just for the tests, which is ugly ;)

In the future it makes anything using this pattern untestable.

Is the `AutorouteEntries` serialized as well, is that why constructor injection is not possible?

If so could we not implement a serialized / deserialized context method, (like an ASP.Net Core `Filter` or `Attribute` that gets it's service provider injected by `CreateInstance`

Reply:
> Is there a way we can avoid having to make these static?

Could be non static

> It is, I imagine, the reason you are having to have special methods, just for the tests, which is ugly ;)

No, this not for this reason, see my above comment.

> Is the AutorouteEntries serialized as well

No

> is that why constructor injection is not possible?

No, here it is just because `AutorouteEntries` is a tenant singleton and `ISession` a scoped service, so we resolve it when executing at the scope level, as we do for `SiteService`.

I will also add a comment in the thread to give more details on `ShellScope`